xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -1840,6 +1800,41 @@ void sendDiscoveryRequest(DiscoveryRequestData request) {
     void sendError(Exception error) {
       requestWriter.onError(error);
     }
+
+    @Override
+    public void onNext(DiscoveryResponse response) {
+      final DiscoveryResponseData responseData = DiscoveryResponseData.fromEnvoyProto(response);
+      if (logger.isLoggable(XdsLogLevel.DEBUG)) {
+        logger.log(XdsLogLevel.DEBUG, "Received {0} response:\n{1}",
+            responseData.getResourceType(), respPrinter.print(response));
+      }
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleDiscoveryResponse(responseData);
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(Status.fromThrowable(t));
+        }
+      });
+    }
```

Comment:
`stub.withWaitForReady().streamAggregatedResources(this)` looks bad. An extreme analogy of this PR:

```java
private final class AdsStreamV3 extends AbstractAdsStream
      implements StreamObserver<DiscoveryResponse>, Runnable {

  @Overrid
  public void run() {
    handleStreamClosed(Status.fromThrowable(t));
  }

  @Override
  public void onError(final Throwable t) {
    syncContext.execute(this);
  }
}
```

Do you think this is clean and elegant?